File: python/sort.py
class Sort:
    def __init__(self):
        pass
        
    def selectionSort(self, arr):
        n = len(arr)
        for i in range(n-1):
            min_index = i
            for j in range(i+1, n):
                if arr[j] < arr[min_index]:
                    min_index = j
            arr[i], arr[min_index] = arr[min_index], arr[i]
        return arr
    
    def InsertionSort(self, arr):
        n = len(arr)
        for i in range(1, n):
            x = arr[i]
            j = i - 1
            while j >= 0 and arr[j] > x:
                arr[j+1] = arr[j]
                j -= 1
            arr[j+1] = x
        return arr
    
    def Partion(self, high, low, arr):
        pivot = arr[high]
        i = low - 1

        for j in range(low, high):
            if arr[j] <= pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]

        arr[i+1], arr[high] = arr[high], arr[i+1]
        return i+1
    
    def quickSort(self, high=None, low=0, arr=[]):
        if high is None:
            high = len(arr) - 1

        if low < high: 
            pivot_index = self.Partion(high, low, arr)
            self.quickSort(pivot_index-1, low, arr)
            self.quickSort(high, pivot_index+1, arr)
        return arr




sort = Sort()  
arr = [64, 56, 32, 7, 0, 23, 45, 6, 11]
print(f"selection sort {sort.selectionSort(arr)}")
print(f"Insertion sort {sort.InsertionSort(arr)}")
print(f"quick sort {sort.quickSort(len(arr)-1, 0, arr)}")
#print(f"Insertion sort {sort.InsertionSort(arr)}")

File: C/sort.c
#include <stdio.h>

void selectionSort(int arr[], int n){

    for (int i = 0; i < n; i++){
        int min_index = i;
        for (int j = i+1; j < n; j++){
            if (arr[j] < arr[min_index]){
                min_index = j;
            }

        }

    int temp = arr[i];
    arr[i] = arr[min_index];
    arr[min_index] = temp;
    }
    
}

void insertionSort(int arr[], int n){
    for (int i = 0; i < n; i++){
        int x = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > x){
            arr[j+1] = arr[j];
            j--;
        }
        arr[j+1] = x;
    }
}

int PrintArray(int *array, int size){
    for (int i = 0; i < size; i++){
        printf("%d ", array[i]);
    }
    printf("\n");
    return 0;
}
 
void swap(int *a, int *b){
    int temp = *a;
    *a = *b;
    *b = temp;
}

int partition(int arr[], int low, int high){
    int pivot = arr[high];
    int i = low -1;

    for (int j = low; j < high; j++){
        if (arr[j] <= pivot){
            i++;
            arr[i], arr[j] = arr[j], arr[j];
        }
    }
    arr[i+1], arr[high] = arr[high], arr[i+1];
    return i+1;
}

int main(){
    int arr[] = {64, 56, 32, 7, 0, 23, 45, 6, 11};
    int n = sizeof(arr) / sizeof(arr[0]);

    selectionSort(arr, n);
    PrintArray(arr, n);

    printf("______\n");

    insertionSort(arr, n);
    PrintArray(arr, n);

    return 0;
}